test/performance/scale_revision_by_load_test.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/knative/test-infra/shared/testgrid"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	qpsPerClient         = 10               // frequencies of requests per client
+	iterationDuration    = 60 * time.Second // iteration duration for a single scale
+	processingTimeMillis = 100              // delay of each request on "server" side
+	targetConcurrency    = 10
+)
+
+var concurrentClients = []int{10, 20, 40, 80, 160, 320}
+
+type scaleEvent struct {
+	oldScale  int
+	newScale  int
+	timestamp time.Time
+}
+
+// TestScaleRevisionByLoad performs several iterations with increasing number of clients
+// while measuring response times, error rates, and time to scale up.
+func TestScaleRevisionByLoad(t *testing.T) {
+	tc := make([]junit.TestCase, 0)
+	for _, numClients := range concurrentClients {
+		t.Run(fmt.Sprintf("clients-%03d", numClients), func(t *testing.T) {
+			tc = append(tc, scaleRevisionByLoad(t, numClients)...)
+		})
+	}
+	if err := testgrid.CreateXMLOutput(tc, t.Name()); err != nil {
+		t.Fatalf("Cannot create output XML: %v", err)
+	}
+}
+
+func scaleRevisionByLoad(t *testing.T, numClients int) []junit.TestCase {
+	perfClients, err := Setup(t)
+	if err != nil {
+		t.Fatalf("Cannot initialize performance client: %v", err)
+	}
+
+	names := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "observed-concurrency",
+	}
+	clients := perfClients.E2EClients
+
+	defer TearDown(perfClients, names, t.Logf)
+	test.CleanupOnInterrupt(func() { TearDown(perfClients, names, t.Logf) })
+
+	t.Log("Creating a new Service")
+	objs, err := test.CreateRunLatestServiceReady(t, clients, &names, &test.Options{
+		ContainerResources: corev1.ResourceRequirements{
+			Requests: corev1.ResourceList{
+				corev1.ResourceCPU:    resource.MustParse("50m"),
+				corev1.ResourceMemory: resource.MustParse("20Mi"),
+			},
+		}},
+		testingv1alpha1.WithConfigAnnotations(map[string]string{"autoscaling.knative.dev/target": fmt.Sprintf("%d", targetConcurrency)}),
+	)
+	if err != nil {
+		t.Fatalf("Failed to create Service: %v", err)
+	}
+
+	domain := objs.Route.Status.Domain
+	endpoint, err := ingress.GetIngressEndpoint(clients.KubeClient.Kube)
+	if err != nil {
+		t.Fatalf("Cannot get service endpoint: %v", err)
+	}
+
+	// Make sure we are ready to serve.
+	st := time.Now()
+	t.Log("Starting to probe the endpoint at", st)
+	_, err = pkgTest.WaitForEndpointState(
+		clients.KubeClient,
+		t.Logf,
+		domain+"/?timeout=10", // To generate any kind of a valid response.
+		test.RetryingRouteInconsistency(pkgTest.IsStatusOK),
+		"WaitForEndpointToServeText",
+		test.ServingFlags.ResolvableDomain)
+	if err != nil {
+		t.Fatalf("The endpoint at domain %s didn't serve the expected response: %v", domain, err)
+	}
+	t.Logf("Took %v for the endpoint to start serving", time.Since(st))
+
+	scaleChannel := make(chan *scaleEvent, 100)
```

Comment:
So why 100? Is that the number of evens we want to record?
Should it dependent on the number of clients?
Some comment would help. Since informer runs a single go routine  if you generate more than 100 events, then event processing thread will block until the test finishes and we start consuming events from this channel.

Reply:
The number was supposed to be big enough to catch any scale events. I've added a new commit that defines a more specific value there, and added a comment.